src/zsim.cpp

```diff
@@ -711,52 +711,116 @@ void VdsoInit() {
 // Register hooks to intercept and virtualize time-related vsyscalls and vdso syscalls, as they do not show up as syscalls!
 // NOTE: getcpu is also a VDSO syscall, but is not patched for now
 
-// Per-thread VDSO data
-struct VdsoPatchData {
+// vDSO patch data for a single thread in a single context (multiple contexts allow signal handlers)
+struct VdsoPatchSingleton {
     // Input arguments --- must save them because they are not caller-saved
     // Careful: REG is 32 bits; PIN_REGISTER, which is the actual type of the
     // pointer, is 64 bits but opaque. We just use ADDRINT, it works
     ADDRINT arg0, arg1;
     VdsoFunc func;
     uint32_t level;  // if 0, invalid. Used for VDSO-internal calls
+
+    VdsoPatchSingleton(ADDRINT _arg0, ADDRINT _arg1, VdsoFunc _func, uint32_t _level) : arg0(_arg0), arg1(_arg1), func(_func), level(_level) {};
+    VdsoPatchSingleton() : arg0(0), arg1(0), func(VdsoFunc(0)), level(0) {};
+};
+
+// Per-thread vDSO context stack. New stack entries are pushed after a pin ContextChange callback (typically a signal),
+// and popped lazily after a return from such a context (a SIGRETURN reason in the ContextChange callback).
+// Without such a mechanism it would be incorrect to call vDSO functions within a signal handler if the prior context were already processing a vDSO
+class VdsoPatchStack {
+    private:
+        std::vector<VdsoPatchSingleton> patchStack[MAX_THREADS];
+
+    public:
+        VdsoPatchStack() {
+            // Each thread starts with vdso info for level 0 (i.e., not in a signal handler)
+            // This is never popped for performance reasons
+            for (uint32_t i = 0; i < MAX_THREADS; i++) {
+                patchStack[i].resize(1);
+            }
+        };
+        VdsoPatchSingleton &peek(THREADID tid) {
+            assert(patchStack[tid].size() > 0);
+            return patchStack[tid].back();
+        };
+        void push(THREADID tid, ADDRINT arg0, ADDRINT arg1, VdsoFunc func, uint32_t level) {
+            patchStack[tid].emplace_back(arg0, arg1, func, level);
+        };
+        void pop(THREADID tid) {
+            assert(patchStack[tid].size() > 0);
+            patchStack[tid].pop_back();
+        };
+        uint32_t size(THREADID tid) {
+            return (uint32_t)patchStack[tid].size();
+        };
 };
-VdsoPatchData vdsoPatchData[MAX_THREADS];
+VdsoPatchStack vdsoPatchData;
+
+// Per-thread signal depth info. (0 -> not in a signal handler, n>0 -> in level 'n' signal handler)
+unsigned int signalStackDepth[MAX_THREADS] = {};
 
 // Analysis functions
 
+// Helper function to laziliy clean up old vdso signal stacks. The alternative would be to register these actions for the SIGRETURN callback
+void vdsoDeferredStackCleanup(THREADID tid) {
+    if ((vdsoPatchData.size(tid) - 1) > signalStackDepth[tid]) {
```

Comment:
Do the depth of vdso calls and `sigStackDepth` need to match? What if there are 4 signal contexts, but just the first and last one called a vdso? Shall we use `while` instead of `if` here?

In general, some comments need to be added to explain the assumed invariance between vdso depth and signal stack depth.